battle-service/src/main/java/com/company/battle/repositories/GameRepository.java

```diff
@@ -1,14 +1,36 @@
 package com.company.battle.repositories;
 
 import com.company.common.models.GameEntity;
+import com.company.common.models.enums.GameVisibility;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
 import org.springframework.stereotype.Repository;
 
+import java.util.List;
 import java.util.Optional;
 import java.util.UUID;
 
 @Repository
 public interface GameRepository extends JpaRepository<GameEntity, UUID> {
 
     Optional<GameEntity> findByLink(String link);
+
+    //ToDo: add sort by date and limit 1
+    @Query(value = "SELECT * FROM game g WHERE g.square_size = :squareSize and g.lines_count_for_win = :linesCount and g.move_time_limit = :timeLimit and g.status = 'PENDING' " +
+            "and g.visibility = 'PUBLIC' and g.o_player_id is not null",
+            nativeQuery = true)
+    Optional<GameEntity> findGameForXPlayer(@Param("squareSize") int squareSize, @Param("linesCount") int linesCountForWin, @Param("timeLimit") int moveTimeLimit);
+
+    @Query(value = "SELECT * FROM game g WHERE g.square_size = :squareSize and g.lines_count_for_win = :linesCount and g.move_time_limit = :timeLimit and g.status = 'PENDING' " +
+            "and g.visibility = 'PUBLIC' and g.x_player_id is not null",
+            nativeQuery = true)
+    Optional<GameEntity> findGameForOPlayer(@Param("squareSize") int squareSize, @Param("linesCount") int linesCountForWin, @Param("timeLimit") int moveTimeLimit);
```

Comment:
it's too hard to follow.
you may use JPA naming to simplify queues https://docs.spring.io/spring-data/jpa/docs/current/reference/html/,
not very clear why you need squareSize,linesCountForWin,moveTimeLimit in all of these queries
findGamebyUserXorByUserO(userId) may return lis of all games in which player particiapated(completed and in progress. you may add additional parameters to filter such games by status.


this is related to all other methods in repositories.
try to segregate queries.

Reply:
These methods are needed in order to find a game with the Pending status and with the same parameters that the player wants to play with. Perhaps you misunderstood the purpose of these methods. This methods search opponents. If you use named methods, the method name will be too long (with setting parameters)